### ISSUE TITLE
jsonrpc sendNotification should return a Promise

### ISSUE THREAD

**Issue by AArnott** [Microsoft/vscode-languageserver-node#99]
The node [stream's write method](https://nodejs.org/api/stream.html#stream_writable_write_chunk_encoding_callback) accepts a callback that is invoked after the written data has been flushed. It also returns a Boolean. Both of these [are ignored](https://github.com/Microsoft/vscode-languageserver-node/blob/master/jsonrpc/src/messageWriter.ts#L82) by the jsonrpc library. If someone were to call `sendNotification` and immediately terminate the process, it might not have been sent to the remote party first. I wonder if it would also fail to send if after calling `sendNotification` the stream were immediately disposed of.

Instead, the callback function should resolve a promise that was returned from the `MessageWriter` and the `sendNotification` method. That way, folks can wait for that to resolve before closing the connection to ensure a reliable transfer.

**Comment by dbaeumer** [comment]
Makes sense however I would not return a promise since the same functionality could be requested for a request (but may be less likely). Need to think about it a little.

**Comment by AArnott** [comment]
Yes, `sendRequest` returns a `Thenable<R>`. I'm proposing that `sendNotification` should return a `Thenable` as well. But where requests resolve when the server responds, notifications should resolve when transmission is complete.
I used the word "Promise" perhaps a bit too loosely.

**Comment by dbaeumer** [comment]
What I was trying to say was (and sorry for not replying for so long): if we add a Thenable to sendNotification which tells that the notification was sent wouldn't someone ask for the same feature on a request. So for requests we have a Thenable which resolve when the message got sent and one that resolves when the response arrived.

**Comment by AArnott** [comment]
Maybe. But I wouldn't be one of them. 
Another similar example is knowing when a response has been fully transmitted from the server side. But no one is asking for that either. It would be rather unnatural from an API standpoint for either of these extra asks, so I wouldn't worry about it unless someone came with a compelling scenario.
